tmva/tmva/inc/TMVA/DNN/CNN/ConvLayer.h

```diff
@@ -145,23 +158,27 @@ class TConvLayer : public VGeneralLayer<Architecture_t> {
-                                       size_t outputNSlices, size_t outputNRows, size_t outputNCols,
-                                       EInitialization init, size_t filterDepth, size_t filterHeight,
-                                       size_t filterWidth, size_t strideRows, size_t strideCols, size_t paddingHeight,
-                                       size_t paddingWidth, Scalar_t dropoutProbability, EActivationFunction f,
-                                       ERegularization reg, Scalar_t weightDecay)
-   : VGeneralLayer<Architecture_t>(batchSize, inputDepth, inputHeight, inputWidth, depth, height, width, 1,
-                                   weightsNRows, weightsNCols, 1, biasesNRows, biasesNCols, outputNSlices, outputNRows,
-                                   outputNCols, init),
-     fFilterDepth(filterDepth), fFilterHeight(filterHeight), fFilterWidth(filterWidth), fStrideRows(strideRows),
+                                       size_t depth, EInitialization init, size_t filterHeight, size_t filterWidth,
+                                       size_t strideRows, size_t strideCols, size_t paddingHeight, size_t paddingWidth,
+                                       Scalar_t dropoutProbability, EActivationFunction f, ERegularization reg,
+                                       Scalar_t weightDecay)
+   : VGeneralLayer<Architecture_t>(batchSize, inputDepth, inputHeight, inputWidth, depth,
+                                   calculateDimension(inputHeight, filterHeight, paddingHeight, strideRows),
+                                   calculateDimension(inputWidth, filterWidth, paddingWidth, strideCols),
+                                   1, depth, calculateNLocalViewPixels(inputDepth, filterHeight, filterWidth),
+                                   1, depth, 1, batchSize, depth,
+                                   calculateNLocalViews(inputHeight, filterHeight, paddingHeight, strideRows,
+                                                        inputWidth, filterWidth, paddingWidth, strideCols),
+                                   init),
+     fFilterDepth(inputDepth), fFilterHeight(filterHeight), fFilterWidth(filterWidth), fStrideRows(strideRows),
      fStrideCols(strideCols), fPaddingHeight(paddingHeight), fPaddingWidth(paddingWidth),
-     fNLocalViewPixels(filterDepth * filterHeight * filterWidth), fNLocalViews(height * width),
+     fNLocalViewPixels(calculateNLocalViewPixels(inputDepth, filterHeight, filterWidth)),
+     fNLocalViews(calculateNLocalViews(inputHeight, filterHeight, paddingHeight, strideRows,
+                                       inputWidth, filterWidth, paddingWidth, strideCols)),
      fDropoutProbability(dropoutProbability), fDerivatives(), fF(f), fReg(reg), fWeightDecay(weightDecay)
 {
-   for (size_t i = 0; i < outputNSlices; i++) {
-      fDerivatives.emplace_back(outputNRows, outputNCols);
+   for (size_t i = 0; i < batchSize; i++) {
+      fDerivatives.emplace_back(depth, fNLocalViews);
```

Comment:
We should have a comment here explaining the meaning of the variables, (i.e. that batchsize is the outputnslices, etc...)

Reply:
I would also request that explanations for e.g. `depth` is provided. What is called depth here is called "number of units" in some literature and "number of channels" in some. Tying our naming convention into what others are using will help people absorb the code quicker!

Reply:
Both valid points, i will address them and update the PR